finder_server.cc

```diff
@@ -0,0 +1,207 @@
+ ServerWriter<ShopInfo>* writer) {
+
+  std::cout << "========== Receving Request Food Id=" << request->food_id() 
+    << " ==========" << std::endl;
+
+  vector<ShopInfo> result = ProcessRequest(request->food_id());
+  long quantity = request->quantity();
+  
+  std::sort(result.begin(), result.end(), Comp());
+
+  for (const auto& info : result){
+    writer->Write(info);
+    quantity -= info.inventory().quantity();
+    if (quantity <= 0)  break;
+  }
+  return Status::OK;
+}
+
+VendorClient::VendorClient(std::shared_ptr<grpc::Channel> vendor_channel):
+                           vendor_stub_(supplyfinder::Vendor::NewStub(vendor_channel)) {}
+
+InventoryInfo VendorClient::InquireInventoryInfo (uint32_t food_id) {
+  FoodID request;
+  request.set_food_id(food_id);
+  InventoryInfo info;
+  ClientContext context;
+  
+  Status status = vendor_stub_->CheckInventory(&context, request, &info);
+  
+  if (status.ok()) {
+    return info;
+  } else {
+    std::cout << status.error_code() << ": " << status.error_message()
+              << std::endl;
+    info.set_price(-1);
+    return info;
+  }
+}
+
+SupplierClient::SupplierClient(std::shared_ptr<grpc::Channel> supplier_channel): 
+                               supplier_stub_(supplyfinder::Supplier::NewStub(supplier_channel)) {}
+
+std::unique_ptr<ClientReader<VendorInfo>>& SupplierClient::InitReader (ClientContext* context_ptr, FoodID& request) {
+  // Release previous reader
+  reader_.reset();
+  // Use the context and request on stack to create new reader
+  // The context and request will be released when ProcessRequest
+  // exits
+  reader_ = supplier_stub_->CheckVendor(context_ptr, request);
+  return reader_;
+}
+
+FinderServiceImpl::FinderServiceImpl(std::string supplier_target_str): 
+                  supplier_client_(grpc::CreateChannel(supplier_target_str, grpc::InsecureChannelCredentials())) {}
+
+void FinderServiceImpl::PrintVendorInfo (const uint32_t id, 
+                                         const VendorInfo& info) {
+  std::cout << "This vendor might have food " << id << std::endl;
+  std::cout << "\tVendor url: " << info.url() << "; name: " << info.name()
+      << "; location: " << info.location() << std::endl;
+}
+
+vector<ShopInfo> FinderServiceImpl::ProcessRequest(uint32_t food_id) {
+  /*
+   * Initiate client with supplier channel. Retrieve a list of vendor address
+   * Then create client
+   * return a vector of <Vendor Info, Inventory Info>
+   */
+  vector<ShopInfo> result;
+  VendorInfo vendor_info;
+  ClientContext context;
+  FoodID request;
+  request.set_food_id(food_id);
+  std::unique_ptr<ClientReader<VendorInfo>>& reader = supplier_client_.InitReader(&context, request);
+  
+  while (reader->Read(&vendor_info)) {
+    // if never connected before, create a new client.
+    FinderServiceImpl::PrintVendorInfo(food_id, vendor_info);
+    const string& url = vendor_info.url();
+    if (vendor_clients_.find(url) == vendor_clients_.end())
```

Comment:
nit: it's more readable to surround conditional blocks with {}, even if it's one line.